File: Rust/2020/05.rs
#![feature(test)]

use itertools::Itertools;

type Input = Vec<u32>;

fn setup(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            line.bytes().fold(0, |acc, c| {
                (acc << 1)
                    | match c {
                        b'F' | b'L' => 0,
                        b'B' | b'R' => 1,
                        _ => panic!(),
                    }
            })
        })
        .collect()
}

fn part1(input: &Input) -> u32 {
    input.iter().copied().max().unwrap()
}

fn part2(input: &Input) -> u32 {
    input
        .iter()
        .sorted_unstable()
        .tuple_windows()
        .find(|(&a, &b)| a + 1 != b)
        .unwrap()
        .0
        + 1
}

aoc::main!(2020, 5);

File: Python/2020/05.py
from lib import *

input = read_input(2020, 5)

lines = input.splitlines()


out = 0
for line in lines:
    row = int(line[:7].replace("F", "0").replace("B", "1"), 2)
    col = int(line[-3:].replace("L", "0").replace("R", "1"), 2)
    out = max(out, row * 8 + col)

print(out)


seats = set()
found = set()
for line in lines:
    row = int(line[:7].replace("F", "0").replace("B", "1"), 2)
    col = int(line[-3:].replace("L", "0").replace("R", "1"), 2)
    found.add(s := row * 8 + col)
    seats |= {s - 1, s + 1}

print(next(iter(seats - found - {min(seats), max(seats)})))
